add swig for tags using Jinja

sanuj opened the pull request:
Working python code:
```
import modshogun as sg
obj = sg.Obj()    # creates CMockObject
obj.getKernel("gaussian")    # returns null
gk = sg.GaussianKernel(5)
obj.setKernel("gaussian", gk)
obj.getKernel("gaussian")    # returns GaussianKernel
```
`shogun-base.i` looks like [this](https://gist.github.com/sanuj/a85986fb35052de761ac251496e90e90).

karlnapf commented on cmake/CommonModularInterface.cmake:
```diff
@@ -4,6 +4,17 @@ if(SYSTEM_INCLUDES)
+    DEPENDS ${COMMON_MODULAR_SRC_DIR}shogun-base.i.py
+    ${COMMON_MODULAR_SRC_DIR}shogun-base.i.jinja2
+    WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
+    COMMENT "Generating shogun-base.i")
```
Cool idea actually

karlnapf commented on src/interfaces/modular/SGBase.i:
```diff
@@ -107,6 +108,7 @@ public void readExternal(java.io.ObjectInput in) throws java.io.IOException, jav
  #include <shogun/base/Version.h>
  #include <shogun/base/Parallel.h>
  #include <shogun/base/SGObject.h>
+ #include <../tests/unit/base/MockObject.h> // Temporary include to test Tags
```
what does "temporary" here mean. That is is removed before a merge?

sanuj replied:
Well, we are not going to keep `CMockObject` in shogun forever. Once shogun classes start using tags-parameter framework, we can remove this. But I think that will take some time, if you want I can remove the "temporary comment.

karlnapf commented on src/interfaces/modular/shogun-base.i.jinja2:
```diff
@@ -0,0 +1,20 @@
+/*
+ * THIS IS A GENERATED FILE!  DO NOT CHANGE THIS FILE!  CHANGE THE
```
why would this be comitted if it is automatically generated?

sanuj replied:
This is the Jinja template for the automatically generated file. This comment is in the template so that it also appears in the auto-generated file. I can also use python code to make this comment in the generated file instead of mentioning it in the template.

lisitsyn replied:
lol

karlnapf commented on src/interfaces/modular/shogun-base.i.py:
```diff
@@ -0,0 +1,51 @@
+#
+# Written (W) 2016 Sanuj Sharma
+
+def entry(templateFile, class_list_file):
```
docstring documentation please

karlnapf commented on src/interfaces/modular/shogun-base.i.py:
```diff
@@ -0,0 +1,51 @@
+    import os
+    basename = os.path.basename(output_file)
+    basename = basename.replace('.cc', '')
+    print "Please install jinja2 for shogun-base.i";
```
Can we please integrate this into cmake properly rather than printing python output and then not failing?

karlnapf commented:
This is cool!
Now we need to tidy it up and integrate into the build properly

sanuj commented:
The travis error is due to this line in `modshogun.R`:
`vaccessors = c('io', 'parallel', 'version', 'm_parameters', 'm_model_selection_parameters', 'm_gradient_parameters', 'm_hash', 'get''get''get''get''get''get''get''get');`

It might be a bug in SWIG:
[swig issue](https://github.com/swig/swig/issues/697)
[swig mailing-list](https://sourceforge.net/p/swig/mailman/message/35071142/)

lisitsyn commented:
Great great

sanuj commented:
@karlnapf @lisitsyn
Updated with following changes:
- Checks of Jinja2 in cmake instead of python script.
- SWIG disabled for `RModular` due to a bug :bug:  in SWIG (see my above comment).
- Auto generated message is now generated by python script.
- Removed temporary comment for `MockObject` in `SGBase.i`.

I think, now we need to worry about [testing it](https://github.com/shogun-toolbox/shogun/issues/3328)  and adding new classes :exclamation:

vigsterkr commented:
@sanuj `RMudular` just works fine: http://buildbot.shogun-toolbox.org/builders/deb3%20-%20modular_interfaces/builds/2933/steps/test%20r%20modular/logs/stdio

if the changes are the one that causes this then your changes are unfortunately not acceptable as it's unacceptable to have RModular disabled.

lisitsyn commented:
@vigsterkr yeah the thing is that we have to wait once it is fixed in swig. This is a PR for feature branch thus I believe it is not really critical (yet).

vigsterkr commented:
@lisitsyn up until it's in a feature branch i dont care... just saying that before merging this anywhere near develop, one should think twice... having RModular lost is a mayor thing.

karlnapf commented:
Lets proceed in feature branch.
But we need to be careful here --- we dont want this to be not merged for months or years, we might need to solve this ourselves.
Do we have a hunch how to do a workaround?

karlnapf commented on src/interfaces/modular/shogun-base.i.py:
```diff
@@ -0,0 +1,48 @@
+"""
+!/usr/bin/env python
+
```
pls not gpl

karlnapf commented on src/interfaces/modular/shogun-base.i.py:
```diff
@@ -0,0 +1,48 @@
+
+    return template.render(templateVars)
+
+# execution
```
can we pls do this properly?

karlnapf commented:
Just checked, can't we provide a workaround? Doesnt seem hard

sanuj commented:
@karlnapf I tried changing the name of `get` to `fetch` and it's working on my local. Can you guys suggest a better name?
Also, I want to be sure and test it with R. What env variable do i need to export to compile with R?

karlnapf commented:
no variable needed.
´-DRModular` is the cmake switch
have a look at the cmakelists.txt for R to see how you can run R examples

vigsterkr commented on src/shogun/lib/config.h.in:
```diff
@@ -41,6 +41,8 @@
 #cmakedefine USE_LOGCACHE 1
 #cmakedefine USE_LOGSUMARRAY 1
 
+#cmakedefine JINJA2_FOUND "@JINJA2_FOUND@"
```
please try to be consequent about naming variables: `USE_*` or `HAVE_*`... and why would you need this macro in C?

vigsterkr commented on CMakeLists.txt:
```diff
@@ -956,6 +956,9 @@ IF(REDUCE_SWIG_DEBUG)
 ENDIF(REDUCE_SWIG_DEBUG)
 
+# For Swig and Unit-tests
+FIND_PACKAGE(Jinja2)
```
this just wrong... why to test it when you dont compile the tests or the modular interfaces?

vigsterkr commented on CMakeLists.txt:
```diff
@@ -957,7 +957,10 @@ IF(REDUCE_SWIG_DEBUG)
-FIND_PACKAGE(Jinja2)
+IF(PythonModular OR LuaModular OR JavaModular OR RubyModular OR
+	OctaveModular OR CSharpModular OR RModular OR PerlModular OR ENABLE_TESTING)
+	FIND_PACKAGE(Jinja2)
```
no need for this. you already have a check for modular to detect swig...

vigsterkr replied:
and please check whether the jinja2 was actually found... and then set a variable.

vigsterkr commented on CMakeLists.txt:
```diff
@@ -79,6 +79,9 @@ INCLUDE_DIRECTORIES(${CMAKE_SOURCE_DIR}/src
 	${CMAKE_BINARY_DIR}/src/shogun)
 
+# detect PYTHON
+FIND_PACKAGE(PythonInterp REQUIRED)
```
the move of this was required because of the `FIND_PACKAGE(Jinja2)`?

vigsterkr replied:
you should rather fix this problem then in `FindJinja2.cmake`, i.e. put in the beginning of that script that for being able to run it you need `FIND_PACKAGE(PythonInterp REQUIRED)`

vigsterkr commented on cmake/CommonModularInterface.cmake:
```diff
@@ -4,6 +4,19 @@ if(SYSTEM_INCLUDES)
+# Generate automatic shogun-base.i from jinja2 templates
+IF(JINJA2_FOUND)
+	ADD_CUSTOM_COMMAND(OUTPUT shogun-base.i
+		COMMAND ${PYTHON_EXECUTABLE} ${COMMON_MODULAR_SRC_DIR}shogun-base.i.py
```
i know that `${COMMON_MODULAR_SRC_DIR}` has a trailing / when it's being set but having `${COMMON_MODULAR_SRC_DIR}shogun-base.i.py` seen makes me feel... mmm so it's not looking good. does the  `${COMMON_MODULAR_SRC_DIR}/shogun-base.i.py` really hurt?

vigsterkr commented on cmake/CommonModularInterface.cmake:
```diff
@@ -32,9 +45,15 @@ FOREACH(file ${MODULAR_FILES})
-	COMMENT "copying SWIG files")
+IF(JINJA2_FOUND)
+	ADD_CUSTOM_TARGET(${MODULAR_NAME}_modular_src
+		DEPENDS shogun ${modular_files} shogun-base.i
```
you could just append `shogun-base.i` to the `${modular_files}`, and call `ADD_CUSTOM_TARGET` once.

vigsterkr commented on examples/undocumented/python_modular/tags_params_modular.py:
```diff
@@ -0,0 +1,67 @@
+    except SystemError as err:
+        print err
+
+    integer = 4
```
couldn't we use any random natural number?

vigsterkr commented on examples/undocumented/python_modular/tags_params_modular.py:
```diff
@@ -0,0 +1,67 @@
+    if integer != obj.getsInt("int") or integer != obj.gets(int_tag):
+        sys.exit(1)
+
+    decimal = 4.4
```
random?

vigsterkr commented on examples/undocumented/python_modular/tags_params_modular.py:
```diff
@@ -0,0 +1,67 @@
+    if kernel != obj.getsKernel("gaussian") or kernel != obj.gets(kernel_tag):
+        sys.exit(1)
+
+    real_vector = RealVector(2)
```
random size

vigsterkr commented on examples/undocumented/python_modular/tags_params_modular.py:
```diff
@@ -0,0 +1,67 @@
+        sys.exit(1)
+
+    real_vector = RealVector(2)
+    real_vector[0] = 0.1
```
random values

vigsterkr commented on src/interfaces/modular/SGBase.i:
```diff
@@ -107,6 +108,7 @@ public void readExternal(java.io.ObjectInput in) throws java.io.IOException, jav
  #include <shogun/base/Version.h>
  #include <shogun/base/Parallel.h>
  #include <shogun/base/SGObject.h>
+ #include <../tests/unit/base/MockObject.h>
```
for relative includes please use `""`, but why do we include `MockObject.h` from (unit) tests into the code itself?

vigsterkr commented on examples/undocumented/python_modular/tags_params_modular.py:
```diff
@@ -0,0 +1,67 @@
+    int_tag = TagInt("int")
+    float_tag = TagFloat("float")
+    kernel_tag = TagKernel("gaussian")
+    vector_tag = TagRealVector("vector")
```
what's `TagRealVector` and `RealVector` are like? and why not the same? :)

vigsterkr commented on src/interfaces/modular/SGBase.i:
```diff
@@ -321,10 +329,13 @@ namespace std {
 %include <shogun/io/SGIO.h>
 %include <shogun/base/Version.h>
 %include <shogun/base/Parallel.h>
+%include <../tests/unit/base/MockObject.h>
```
this is still included without guards for testing...
and could you please use `#include "..."` instead of doing some magical relative path hacking for `#include <>`. note: read the manual of header include

vigsterkr commented on src/interfaces/modular/SGBase.i:
```diff
@@ -108,6 +109,10 @@ public void readExternal(java.io.ObjectInput in) throws java.io.IOException, jav
  #include <shogun/base/SGObject.h>
 
+#ifdef ENABLE_TESTING
+ #include "../tests/unit/base/MockObject.h"
```
does this actually work?
